src/uopool/memory.rs

```diff
@@ -0,0 +1,352 @@
+
+use super::{Mempool, MempoolId};
+
+pub type UserOperationsBySender = HashMap<Address, HashSet<UserOperationHash>>;
+
+#[derive(Educe)]
+#[educe(Debug)]
+pub struct MemoryMempool {
+    user_operations: Arc<RwLock<HashMap<UserOperationHash, UserOperation>>>, // user_operation_hash -> user_operation
+    user_operations_by_entry_point: Arc<RwLock<HashMap<MempoolId, HashSet<UserOperationHash>>>>, // mempool_id -> user_operations
+    user_operations_by_sender: Arc<RwLock<HashMap<MempoolId, UserOperationsBySender>>>, // mempool_id -> sender -> user_operations
+}
+
+impl Mempool for MemoryMempool {
+    fn add(
+        &mut self,
+        user_operation: UserOperation,
+        entry_point: Address,
+        chain_id: U256,
+    ) -> anyhow::Result<UserOperationHash> {
+        let hash = user_operation.hash(entry_point, chain_id);
+
+        // replace user operation
+        if self.user_operations.read().contains_key(&hash) {
+            let user_operations = self.user_operations.read();
+            let user_operation_old = user_operations.get(&hash).unwrap();
+            let max_priority_fee_per_gas_diff = user_operation.max_priority_fee_per_gas
+                - user_operation_old.max_priority_fee_per_gas;
+            let max_fee_per_gas_diff =
+                user_operation.max_fee_per_gas - user_operation_old.max_fee_per_gas;
+            if !(user_operation.sender == user_operation_old.sender
+                && user_operation.nonce == user_operation_old.nonce
+                && user_operation.max_priority_fee_per_gas
+                    > user_operation_old.max_priority_fee_per_gas
+                && max_priority_fee_per_gas_diff == max_fee_per_gas_diff)
+            {
+                return Err(anyhow::anyhow!("User operation already exists"));
+            }
+        }
+
+        let mut user_operations = self.user_operations.write();
+        let mut user_operations_by_entry_point = self.user_operations_by_entry_point.write();
+        let mut user_operations_by_sender = self.user_operations_by_sender.write();
+
+        let id = MemoryMempool::id(entry_point, chain_id);
+        user_operations_by_entry_point
+            .get_mut(&id)
+            .unwrap()
+            .insert(hash);
+        user_operations_by_sender
+            .get_mut(&id)
+            .unwrap()
+            .entry(user_operation.sender)
+            .or_insert_with(Default::default)
+            .insert(hash);
+        user_operations.insert(hash, user_operation);
+
+        Ok(hash)
+    }
+
+    fn get(&self, user_operation_hash: UserOperationHash) -> anyhow::Result<UserOperation> {
+        if !self
+            .user_operations
+            .read()
+            .contains_key(&user_operation_hash)
+        {
+            return Err(anyhow::anyhow!("User operation not found"));
+        }
+        Ok(self
+            .user_operations
+            .read()
+            .get(&user_operation_hash)
+            .unwrap()
```

Comment:
Unsafe unwrap here. This function return `Result`. It would be better turns it into Err